### ISSUE TITLE
Propagate seen aggregate attestations

### ISSUE THREAD

**Issue by AgeManning** [ethereum/eth2.0-specs#2183]
Currently, the gossipsub validation rules for the `beacon_aggregate_and_proof` topic specify that a client should `IGNORE` messages that contain an aggregate that has previously been seen. 

> [IGNORE] The valid aggregate attestation defined by hash_tree_root(aggregate) has not already been seen (via aggregate gossip, within a verified block, or through the creation of an equivalent aggregate locally).

This gives rise to a few issues within the gossipsub protocol. 

There are on average, `TARGET_AGGREGATORS_PER_COMMITTEE` (16 for mainnet) aggregators per committee. These aggregators will often agree on the same aggregate attestation and submit individual `SignedAggregateAndProof` objects to the `beacon_aggregate_and_proof` topic. These messages are unique (as the aggregators and hence their signatures are unique) and therefore have their own unique gossipsub message-ids.

Therefore the standard operation of a node is to receive one valid aggregate and then `IGNORE` (aka not forward, but still gossip about) any remaining duplicate aggregate attestations as specified by the topic rule mentioned above. Different nodes on the network may receive different aggregates as their first and `IGNORE` others. The gossipsub protocol will then gossip about the extra message-ids that were not forwarded. Other nodes will then notice they didn't receive these extra aggregates (gossipsub message ids) and request the unknown gossip-ids via gossipsub IWANT messages from their peers. Eventually all the aggregates should still propagate to all nodes in the network, albeit inefficiently. 

Besides the inefficient propagation of duplicate aggregate attestations, peers may also get penalized for not propagating all the messages on the topic (i.e it appears some nodes are censoring some message-ids as they don't forward them). 

An easy solution to handling the scoring issue and inefficiencies of propagating via IHAVE/IWANT gossip, is to relax the topic requirement such that all nodes still forward duplicate aggregate attestations, provided the signed aggregate has not been seen before. This will make the propagation more efficient and reduce the messages being sent and received via IHAVE/IWANT. 
Potentially in the future, we may want to think about a way to filter out the duplicate aggregate attestations to save bandwidth, but this may involve complicated changes to the message-id function and is something we can address in the future.

**Comment by djrtwo** [comment]
Adding this to HF1

**Comment by ralexstokes** [comment]
Making a note the outcome here may be relevant for the sync committee aggregation scheme we are working on in https://github.com/ethereum/eth2.0-specs/pull/2201

**Comment by AgeManning** [comment]
This may require further discussion modification after analysing the impact of the proposed change here. 

This issue was originally created from the perspective of making our gossipsub layer as efficient as we can, but the assumption here is that we want all messages that we publish on a topic to be propagated to all peers subscribed to the topic (how gossipsub is designed to function). It seems there may be an exception for the `beacon_aggregate_and_proof` topic. 

In the specs, we choose an average of 16 aggregators to send aggregated attestations on the `beacon_aggregate_and_proof` topic. If the aggregates are identical, then we have about a 16x message redundancy. Before this issue, clients would `IGNORE` (in the gossipsub validation sense) duplicates. This would be a race condition as to which aggregate was sent to a node first. If one aggregate made it to all nodes in the network first, then all other published aggregates would be ignored by their recpient peers and not propagate through the network. If there is a mix, i.e some nodes recieved one aggregate first then another second, both would propagate slowly and not fully via IHAVE/IWANT from some nodes that consider one aggregate valid over another (based on the time they received them). The end result is we were reducing redundancy (the amount of which is based on timing) in the aggregate attestations that are published and propagated through the network.

As can be seen from https://github.com/sigp/lighthouse/pull/2750 and https://github.com/prysmaticlabs/prysm/issues/9935 the changes orginially suggested in this issue, now fully propagate the redundant messages which appear to increase bandwidth and also increase CPU-time as we have to still verify the redundant aggregate signatures. 

Although our previous approach does reduce the redundancy, somewhat ad-hoc (based on propagation timing etc), the way we are doing it is probably not the best way as gossipsub was not desigend for this partial propagation and we still have some wasted messages in IHAVE/IWANT depending on the timing and propagation of published attestations.

A few things we could do (from easy to hard):
- Leave this as it was and don't make the modifications (i.e revert the PR made becase of this issue)
- Set validators to not publish an aggregate if they have already seen an identical aggregate (may help a little, maybe even somehow order aggregator publishing time)
- Make the message-id of the `beacon_aggregate_and_proof` topic such that identical aggregates (although from different aggregators) are identical. (This is  would leverage gossipsub internals to handle the duplicates for us before reaching the application. At the network level, this would make aggregates from different validators interchangeable (or look identical), which I dont see would be an immediate issue. The problem here is that we all have to modify gossipsub to allow for a message-id per topic, and potentially the bandwidth saving is not worth the engineering effort. 
- Some other spec-level modification to reduce duplication.

**Comment by nisdas** [comment]
I dont see any easy way to make `gossipsub` aware of the intricacies of how `attestation_aggregates` work in practice as there doesn't seem to be an easy way to handle partial propagation.  We might have to accept the inefficiency that comes with the excessive `IWANT` requests.

> Set validators to not publish an aggregate if they have already seen an identical aggregate (may help a little, maybe even somehow order/randomize aggregator publishing time)

This would be very tight, assuming all validators in the network are synced within 500ms of each other, packing in the aggregates and broadcasting them would take potentially 500ms - 1000ms. Relying on a node to receive an aggregate before broadcasting its own would have a limited impact on duplicates being propagated.

> Leave this as it was and don't make the modifications (i.e revert the PR made becase of this issue)

I think this would be the best course of action for now.

**Comment by djrtwo** [comment]
The timing delay mechanism is a nice exploration that we can/should simulate, but I do agree with @nisdas that a reversion might be the most reasonable course of action in the immediate term

**Comment by dapplion** [comment]
I think this is a great opportunity to do a post mortem for this incident to clearly define and structure the research and experimentation required before merging changes to the spec.

**Comment by potuz** [comment]
> I think this is a great opportunity to do a post mortem for this incident to clearly define and structure the research and experimentation required before merging changes to the spec.

Specially because the issue at hand has an explicit comment from Proto regarding a tradeoff with bandwidth, so this wasn't actually quantified before it was merged: 
<https://github.com/ethereum/consensus-specs/pull/2225#pullrequestreview-611747244>

**Comment by djrtwo** [comment]
Not to mention -- the original mechanism did *not* work as we intended (strictly dropping dupes) and was short-circuited  by IWANT/IHAVE. So it kind of worked in practice but by accident. So this was an attempt to fix an issue, not to optimize, but any change in network stack should likely have a resource consumption analysis at this point because it can greatly affect end users

**Comment by AgeManning** [comment]
Yep this wasn't supposed to be strictly an optimization. It was correcting the spec to function as we intended (every message gets propagated to everyone on a topic). Turns out the way we were doing it before, has some positives with resource usage. The downside still is that messages published on the `beacon_aggreate_and_proof` topic can be partially propagated. So some nodes on the network can receive different messages than others even though they are subscribed to the same topic. It turns out this doesn't seem to harm anything and costs us less resources.
